File: clients/php/contract_test.php
<?php
// Runs the shared SDK contract scenario against real compiled binaries.
//
// Every SESAME SDK drives this same sequence, so a divergence in any one
// client fails its own suite rather than hiding behind a mock.

declare(strict_types=1);

require __DIR__ . '/sesame.php';

use Sesame\Client;
use Sesame\ProtocolError;

$checks = 0;

function areEqual(mixed $expected, mixed $actual, string $what): void
{
    global $checks;
    $checks++;
    if ($expected !== $actual) {
        fwrite(STDERR, sprintf("%s: expected %s, got %s\n", $what,
            var_export($expected, true), var_export($actual, true)));
        exit(1);
    }
}

function codeOf(callable $action): string
{
    try {
        $action();
    } catch (ProtocolError $error) {
        return $error->errorCode;
    }
    fwrite(STDERR, "expected a protocol error\n");
    exit(1);
}

function repositoryRoot(): string
{
    $candidate = __DIR__;
    while ($candidate !== '/' && $candidate !== '') {
        if (is_file($candidate . '/go.mod')) {
            return $candidate;
        }
        $candidate = dirname($candidate);
    }
    fwrite(STDERR, "could not locate the repository root\n");
    exit(1);
}

function build(string $root, string $workspace, string $name, string $source): string
{
    $output = $workspace . '/' . $name;
    $command = sprintf(
        'cd %s && CGO_ENABLED=0 GOTOOLCHAIN=auto go build -trimpath -o %s %s',
        escapeshellarg($root), escapeshellarg($output), escapeshellarg($source),
    );
    exec($command, $ignored, $status);
    if ($status !== 0) {
        fwrite(STDERR, "go build {$source} failed\n");
        exit(1);
    }
    return $output;
}

$root = repositoryRoot();
$workspace = sys_get_temp_dir() . '/sesame-php-contract-' . getmypid();
mkdir($workspace, 0o700, true);

try {
    $sesameBinary = build($root, $workspace, 'sesame', './cmd/sesame');
    $fakeFylo = build($root, $workspace, 'fake-fylo', './internal/adapters/fylo/testdata/fakefylo');
    $fyloRoot = $workspace . '/root';
    mkdir($fyloRoot, 0o700);

    $client = new Client($sesameBinary, null, $fakeFylo, $fyloRoot);

    // System operations report a storage-backed process.
    areEqual('ok', $client->ping()['status'], 'ping status');
    areEqual('ok', $client->readiness()['status'], 'readiness status');
    areEqual('sesame', $client->version()['name'], 'version name');
    areEqual(true, $client->metrics()['storage_configured'], 'storage configured');

    // Tenant and principal.
    $tenantId = $client->tenantBootstrap('acme')['tenant']['tenant_id'];
    $principal = $client->principalCreate($tenantId, 'human', 'email', 'Alice@Example.com');
    $principalId = $principal['principal_id'];
    areEqual('alice@example.com', $principal['identifier']['value'],
        'the engine normalizes the identifier');

    // A duplicate claim returns the stable conflict code.
    areEqual('identifier_conflict', codeOf(fn() =>
        $client->principalCreate($tenantId, 'workload', 'email', 'alice@example.com')),
        'duplicate identifier');

    // Authorization: deny by default, allow after a grant, deny after revoke.
    $roleId = $client->roleCreate($tenantId, 'reader', [
        ['action' => 'doc:read', 'resource' => 'project:*'],
    ])['role_id'];

    $denied = $client->decide($tenantId, $principalId, 'doc:read', 'project:alpha');
    areEqual('deny', $denied['decision'], 'pre-grant decision');
    areEqual('deny_no_grant', $denied['reason_code'], 'pre-grant reason');

    $grantId = $client->grantCreate($tenantId, $principalId, $roleId)['grant_id'];
    $allowed = $client->decide($tenantId, $principalId, 'doc:read', 'project:alpha');
    areEqual('allow', $allowed['decision'], 'post-grant decision');
    areEqual('allow_role_grant', $allowed['reason_code'], 'post-grant reason');

    $client->grantRevoke($grantId);
    areEqual('deny', $client->decide($tenantId, $principalId, 'doc:read', 'project:alpha')['decision'],
        'post-revoke decision');

    // Authentication: password login, session verification, revocation.
    $password = 'correct horse battery staple';
    $client->setPassword($principalId, $password);
    $begun = $client->authnBegin($tenantId, 'email', 'Alice@Example.com');
    areEqual('awaiting_factor', $begun['state'], 'begin state');
    $transactionId = $begun['transaction_id'];

    $wrong = $client->authnVerifyPassword($transactionId, 'wrong password value');
    areEqual(4, $wrong['attempts_left'], 'attempts after a wrong password');

    $verified = $client->authnVerifyPassword($transactionId, $password);
    areEqual('password', $verified['assurance'], 'assurance');

    $issued = $client->authnComplete($transactionId, 3600);
    $sessionId = $issued['session_id'];
    $secret = $issued['session_secret'];

    $session = $client->sessionVerify($sessionId, $secret);
    areEqual($principalId, $session['principal_id'], 'verified session principal');
    areEqual(false, array_key_exists('secret_digest', $session),
        'the stored digest must not cross the boundary');

    areEqual('session_not_found', codeOf(fn() => $client->sessionVerify($sessionId, 'nope')),
        'wrong session secret');
    $client->sessionRevoke($sessionId, 'test');
    areEqual('session_inactive', codeOf(fn() => $client->sessionVerify($sessionId, $secret)),
        'revoked session');

    // An unknown identifier is indistinguishable from a known one.
    $unknown = $client->authnBegin($tenantId, 'email', 'ghost@example.com');
    areEqual($begun['state'], $unknown['state'], 'unknown identifier state');
    $attempt = $client->authnVerifyPassword($unknown['transaction_id'], $password);
    areEqual(4, $attempt['attempts_left'], 'unknown identifier attempts');
    areEqual(false, array_key_exists('assurance', $attempt), 'unknown identifier assurance');

    // Unknown records return stable codes rather than transport failures.
    areEqual('tenant_not_found', codeOf(fn() => $client->tenantGetByName('missing')), 'missing tenant');
    areEqual('operation_not_found', codeOf(fn() => $client->request('identity.unknown')),
        'unknown operation');

    $client->close();
    printf("ok\tphp contract scenario\t%d checks\n", $checks);
} finally {
    exec('rm -rf ' . escapeshellarg($workspace));
}

File: clients/ruby/contract_test.rb
# frozen_string_literal: true

# Runs the shared SDK contract scenario against real compiled binaries.
#
# Every SESAME SDK drives this same sequence, so a divergence in any one
# client fails its own suite rather than hiding behind a mock.

require 'fileutils'
require 'tmpdir'
require_relative 'sesame'

CHECKS = { count: 0 }

def are_equal(expected, actual, what)
  CHECKS[:count] += 1
  return if expected == actual

  warn("#{what}: expected #{expected.inspect}, got #{actual.inspect}")
  exit(1)
end

def code_of
  yield
  warn('expected a protocol error')
  exit(1)
rescue Sesame::ProtocolError => error
  error.code
end

def repository_root
  candidate = __dir__
  while candidate != '/' && !candidate.empty?
    return candidate if File.file?(File.join(candidate, 'go.mod'))

    candidate = File.dirname(candidate)
  end
  warn('could not locate the repository root')
  exit(1)
end

def build(root, workspace, name, source)
  output = File.join(workspace, name)
  system(
    { 'CGO_ENABLED' => '0', 'GOTOOLCHAIN' => 'auto' },
    'go', 'build', '-trimpath', '-o', output, source,
    chdir: root
  ) || (warn("go build #{source} failed") || exit(1))
  output
end

root = repository_root
workspace = Dir.mktmpdir('sesame-ruby-contract-')
begin
  sesame_binary = build(root, workspace, 'sesame', './cmd/sesame')
  fake_fylo = build(root, workspace, 'fake-fylo', './internal/adapters/fylo/testdata/fakefylo')
  fylo_root = File.join(workspace, 'root')
  FileUtils.mkdir_p(fylo_root, mode: 0o700)

  client = Sesame::Client.new(binary: sesame_binary, fylo_binary: fake_fylo, fylo_root: fylo_root)

  # System operations report a storage-backed process.
  are_equal('ok', client.ping['status'], 'ping status')
  are_equal('ok', client.readiness['status'], 'readiness status')
  are_equal('sesame', client.version['name'], 'version name')
  are_equal(true, client.metrics['storage_configured'], 'storage configured')

  # Standards dispatch pins contract v1 and rejects boundary injection.
  standards = client.standards_dispatch(
    'contract_version' => 'unsupported',
    'endpoint' => 'oidc.token',
    'method' => 'GET'
  )
  are_equal('1', standards['contract_version'], 'standards contract version')
  are_equal(405, standards['status'], 'standards method status')
  are_equal('POST', standards.dig('headers', 'allow'), 'standards allowed method')
  are_equal('application/json', standards.dig('headers', 'content-type'),
            'standards content type')
  are_equal({ 'error' => 'invalid_request' }, standards['body'], 'standards error body')
  are_equal('invalid_request', code_of do
    client.standards_dispatch(
      'contract_version' => 'unsupported',
      'endpoint' => 'oidc.token',
      'method' => 'POST',
      'authorization' => "Basic safe\r\nX-Injected: yes"
    )
  end, 'standards boundary injection')

  # Tenant and principal.
  tenant_id = client.tenant_bootstrap('acme')['tenant']['tenant_id']
  principal = client.principal_create(tenant_id, 'human', 'email', 'Alice@Example.com')
  principal_id = principal['principal_id']
  are_equal('alice@example.com', principal['identifier']['value'],
            'the engine normalizes the identifier')

  # A duplicate claim returns the stable conflict code.
  are_equal('identifier_conflict',
            code_of { client.principal_create(tenant_id, 'workload', 'email', 'alice@example.com') },
            'duplicate identifier')

  # Authorization: deny by default, allow after a grant, deny after revoke.
  role_id = client.role_create(tenant_id, 'reader',
                               [{ action: 'doc:read', resource: 'project:*' }])['role_id']

  denied = client.decide(tenant_id, principal_id, 'doc:read', 'project:alpha')
  are_equal('deny', denied['decision'], 'pre-grant decision')
  are_equal('deny_no_grant', denied['reason_code'], 'pre-grant reason')

  grant_id = client.grant_create(tenant_id, principal_id, role_id)['grant_id']
  allowed = client.decide(tenant_id, principal_id, 'doc:read', 'project:alpha')
  are_equal('allow', allowed['decision'], 'post-grant decision')
  are_equal('allow_role_grant', allowed['reason_code'], 'post-grant reason')

  client.grant_revoke(grant_id)
  are_equal('deny', client.decide(tenant_id, principal_id, 'doc:read', 'project:alpha')['decision'],
            'post-revoke decision')

  # Authentication: password login, session verification, revocation.
  password = 'correct horse battery staple'
  client.set_password(principal_id, password)
  begun = client.authn_begin(tenant_id, 'email', 'Alice@Example.com')
  are_equal('awaiting_factor', begun['state'], 'begin state')
  transaction_id = begun['transaction_id']

  wrong = client.authn_verify_password(transaction_id, 'wrong password value')
  are_equal(4, wrong['attempts_left'], 'attempts after a wrong password')

  verified = client.authn_verify_password(transaction_id, password)
  are_equal('password', verified['assurance'], 'assurance')

  issued = client.authn_complete(transaction_id, 3600)
  session_id = issued['session_id']
  secret = issued['session_secret']

  session = client.session_verify(session_id, secret)
  are_equal(principal_id, session['principal_id'], 'verified session principal')
  are_equal(false, session.key?('secret_digest'),
            'the stored digest must not cross the boundary')

  are_equal('session_not_found', code_of { client.session_verify(session_id, 'nope') },
            'wrong session secret')
  client.session_revoke(session_id, 'test')
  are_equal('session_inactive', code_of { client.session_verify(session_id, secret) },
            'revoked session')

  # An unknown identifier is indistinguishable from a known one.
  unknown = client.authn_begin(tenant_id, 'email', 'ghost@example.com')
  are_equal(begun['state'], unknown['state'], 'unknown identifier state')
  attempt = client.authn_verify_password(unknown['transaction_id'], password)
  are_equal(4, attempt['attempts_left'], 'unknown identifier attempts')
  are_equal(false, attempt.key?('assurance'), 'unknown identifier assurance')

  # Unknown records return stable codes rather than transport failures.
  are_equal('tenant_not_found', code_of { client.tenant_get_by_name('missing') }, 'missing tenant')
  are_equal('operation_not_found', code_of { client.request('identity.unknown') },
            'unknown operation')

  client.close
  puts "ok\truby contract scenario\t#{CHECKS[:count]} checks"
ensure
  FileUtils.remove_entry(workspace, true)
end
